library/modules/Items.cpp

```diff
@@ -2270,6 +2270,74 @@ bool Items::markForTrade(df::item *item, df::building_tradedepotst *depot) {
     return true;
 }
 
+bool Items::canMelt(df::item *item)
+{
+    CHECK_NULL_POINTER(item);
+
+    switch(item->getType()) {
+        // these are not meltable, even if made from metal
+        case df::item_type::CORPSE:
+        case df::item_type::CORPSEPIECE:
+        case df::item_type::REMAINS:
+        case df::item_type::FISH:
+        case df::item_type::FISH_RAW:
+        case df::item_type::VERMIN:
+        case df::item_type::PET:
+        case df::item_type::FOOD:
+        case df::item_type::EGG:
+        case df::item_type::BAR:
+            return false;
+        default:
+            break;
+    }
+
+    // original check from the logistics module
+    MaterialInfo mat(item);
+    if (mat.getCraftClass() != df::craft_material_class::Metal)
+        return false;
+
+    for (auto &g : item->general_refs) {
+        switch (g->getType()) {
+            case df::general_ref_type::CONTAINS_ITEM:
+            case df::general_ref_type::UNIT_HOLDER:
+            case df::general_ref_type::CONTAINS_UNIT:
+                return false;
+            case df::general_ref_type::CONTAINED_IN_ITEM:
+            {
+                df::item *c = g->getItem();
+                for (auto &gg : c->general_refs) {
+                    if (gg->getType() == df::general_ref_type::UNIT_HOLDER)
+                        return false;
+                }
+                break;
+            }
+            default:
+                break;
+        }
+    }
+
+    if (item->flags.bits.artifact)
+        return false;
+
+    return true;
+};
+
+bool Items::markForMelting(df::item *item)
```

Comment:
this function should guard against designating an item for melt that is already designated for melt, which could result in the item being the vector twice, which is likely bad

Reply:
I believe `insert_into_vector` is a noop if the item already exists, so it should be safe. The not-at-all-similarly-named-and-absolutely-not-confusing `vector_insert_at` method has no such protections and would insert blindly.

Reply:
Right. `insert_into_vector` checks if the object is present, then calls `vector_insert_at`. https://github.com/DFHack/dfhack/blob/bbc79d70c04a993d84da591e8cd0bf18bab15075/library/include/MiscUtils.h#L215-L235